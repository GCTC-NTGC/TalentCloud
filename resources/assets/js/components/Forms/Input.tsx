import React from "react";

export interface InputProps {
  htmlId: string;
  name: string;
  label: string;
  required?: boolean;
  placeholder?: string;
  checked?: boolean;
  type?: string;
  inputType?: string;
  minLength?: number;
  maxLength?: number;
  value?: string | number;
  errorText?: string;
  grid?: string;
  onChange: (event: React.ChangeEvent<HTMLInputElement>) => void;
  onBlur?: (event: React.ChangeEvent<HTMLInputElement>) => void;

  // formik
  field?: any;
  form?: any;

  // radio context id
  contextId?: string;
  trigger?: boolean;
}

const Input: React.FunctionComponent<InputProps> = ({
  htmlId,
  name,
  label,
  required,
  placeholder,
  checked,
  type,
  inputType,
  value,
  onChange,
  errorText,
  grid,
  minLength,
  maxLength,
  onBlur,
  field,
  form: { errors, touched },
  contextId,
  trigger,
}): React.ReactElement => {
  function renderCheckbox(): React.ReactElement {
    return (
      <div data-c-grid-item={grid}>
        <label htmlFor={htmlId}>
          <input
            id={htmlId}
            name={field.name || name}
            type="checkbox"
            checked={checked}
            value={value}
            onChange={field.onChange || onChange}
            onBlur={field.onBlur || onBlur}
            required={required}
          />
          <span>{label}</span>
        </label>
      </div>
    );
  }

  function renderRadio(): React.ReactElement {
    const clicked: boolean = htmlId === field.value;
    return (
      <label
        htmlFor={htmlId}
        data-tc-wenv-id={contextId}
        data-tc-wenv-trigger={trigger}
        className={clicked ? "active" : ""}
      >
        <input
          data-c-font-weight="800"
          id={htmlId}
          name={field.name}
          type="radio"
          checked={clicked}
          value={htmlId}
          onChange={field.onChange}
          onBlur={field.onBlur}
        />
        <span>{label}</span>
      </label>
    );
  }

  function renderText(): React.ReactElement {
    return (
      <div
        data-c-input={type || "text"}
        data-c-grid-item={grid}
        data-c-required={required}
        data-c-invalid={touched[field.name] && errors[field.name] ? true : null}
      >
        <label htmlFor={htmlId}>{label}</label>
        {required && <span>Required</span>}
        <div>
          <input
            id={htmlId}
            name={field.name || name}
            placeholder={placeholder}
            type={type || "text"}
            value={field.value || value}
            onChange={field.onChange || onChange}
            minLength={minLength}
            maxLength={maxLength}
            onBlur={field.onBlur || onBlur}
          />
        </div>
        <span>
          {(touched[field.name] && errors[field.name]) ||
            errorText ||
            "Something went wrong."}
        </span>
      </div>
    );
  }

  switch (inputType || type) {
    case "radio":
      return renderRadio();
    case "checkbox":
      return renderCheckbox();
    default:
      return renderText();
  }
};

export default Input;
