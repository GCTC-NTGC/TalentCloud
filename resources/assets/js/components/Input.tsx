import React from "react";
import { FormattedMessage } from "react-intl";
import { inputMessages } from "./Form/Messages";

export interface InputProps {
  /** HTML id of the input element */
  id: string;
  /** HTML name of the input element */
  name: string;
  /** Holds text for label associated with input element */
  label: string;
  /** boolean indicating if input must have a value, or not */
  required?: boolean;
  /** Boolean that sets the select input to invalid */
  invalid?: boolean | null;
  /** Let's you specify example text that appears in input element when empty */
  placeholder?: string;
  /**
   * The input type: https://developer.mozilla.org/en-US/docs/Web/HTML/Element/input#Form_%3Cinput%3E_types
   * Defaults to 'text'
   * */
  type?: string;
  /** Minimum length of characters the text value can be */
  minLength?: number;
  /** Maximum length of characters the text value can be */
  maxLength?: number;
  /** Minimum value for the input */
  min?: number;
  /** Maximum value for the input */
  max?: number;
  /** The value of the input */
  value?: string | number | string[];
  /** Error text that appers underneath if error occurs (eg. required) */
  errorText?: string;
  /** data-clone-grid-item value: https://designwithclone.ca/#flexbox-grid */
  grid?: string;
  /** Event listener which fires when a change event occurs (varies on input type) */
  onChange: (event: React.ChangeEvent<HTMLInputElement>) => void;
  /** Event listener which fires when a input loses focus */
  onBlur?: (event: React.ChangeEvent<HTMLInputElement>) => void;
}

const Input: React.FunctionComponent<InputProps> = ({
  id,
  name,
  label,
  required,
  placeholder,
  type,
  value,
  invalid,
  errorText,
  grid,
  minLength,
  maxLength,
  min,
  max,
  onBlur,
  onChange,
}): React.ReactElement => (
  <div
    data-c-input={type || "text"}
    data-c-grid-item={grid}
    data-c-required={required || null}
    data-c-invalid={invalid || null}
  >
    <label htmlFor={id}>{label}</label>
    <span>
      <FormattedMessage {...inputMessages.required} />
    </span>
    <div>
      <input
        id={id}
        name={name}
        required={required}
        placeholder={placeholder}
        type={type || "text"}
        value={value}
        minLength={minLength}
        maxLength={maxLength}
        min={min}
        max={max}
        onChange={onChange}
        onBlur={onBlur}
      />
    </div>
    <span>{errorText || <FormattedMessage {...inputMessages.error} />}</span>
  </div>
);

export default Input;
