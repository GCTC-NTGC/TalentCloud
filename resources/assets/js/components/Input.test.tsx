import React from "react";
import renderer from "react-test-renderer";
import Input from "./Input";

it("renders correctly", (): void => {
  const tree = renderer
    .create(
      <Input
        id="sample-input"
        name="name"
        label="label"
        required={false}
        placeholder="Write stuff"
        type="text"
        minLength={0}
        maxLength={30}
        value="Hello World"
        onChange={(): void => {}}
      />,
    )
    .toJSON();
  expect(tree).toMatchSnapshot();
});
