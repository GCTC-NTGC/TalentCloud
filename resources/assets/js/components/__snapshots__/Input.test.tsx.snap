// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders correctly 1`] = `
<div
  data-c-input="text"
>
  <label
    htmlFor="sample-input"
  >
    label
  </label>
  <div>
    <input
      data-c-font-weight="800"
      id="sample-input"
      maxLength={30}
      minLength={0}
      name="name"
      onChange={[Function]}
      placeholder="Write stuff"
      type="text"
      value="Hello World"
    />
  </div>
  <span>
    Something went wrong.
  </span>
</div>
`;
